File: RasPISamples/src/C/system.c
/*
 * Compile with
 * gcc -o system system.c
 ***********************************************************************
 * Oliv proudly did it.
 * 
 * TODO Free resources.
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <float.h>
#include <string.h>

#define TRUE 1
#define FALSE 0

#define DEBUG FALSE

double * minor(int dim, double * matrix, int row, int col) {
	double * small = (double *) malloc((dim - 1) * (dim - 1) * sizeof(double));
	for (int c=0; c<dim; c++) {
		if (c != col) {
			for (int r=0; r<dim; r++) {
				if (r != row) {
					small[(((r < row) ? r : (r - 1)) * (dim - 1)) + ((c < col) ? c : (c - 1))] = matrix[(r * dim) + c];
				}
			}
		}
	}
	return small;
}

double determinant(int dim, double * matrix) {
	double v = 0;
	if (dim == 1) {
		v = matrix[0];
	} else {
		// C: column in major
		for (int C=0; C<dim; C++) { // Walk thru first line
			double minDet = determinant(dim - 1, minor(dim, matrix, 0, C));
			v += (matrix[C] * minDet * pow(-1, C + 1 + 1));
		}
	}
	return v;
}

double * comatrix(int dim, double * matrix) {
	double * comat = (double *) malloc(dim * dim * sizeof(double));
	for (int r=0; r<dim; r++) {
		for (int c=0; c<dim; c++) {
			comat[(r * dim) + c] = determinant((dim - 1), minor(dim, matrix, r, c)) * pow(-1, (r + c + 2));
		}
	}
	return comat;
}

double * transposed(int dim, double * matrix) {
	double * transp = (double *) malloc(dim * dim * sizeof(double));
	for (int r=0; r<dim; r++) {
		for (int c=0; c<dim; c++) {
			transp[(r * dim) + c] = matrix[(c * dim) + r];
		}
	}
	return transp;
}

double * multiply(int dim, double * matrix, double n) {
	double * mult = (double *) malloc(dim * dim * sizeof(double));
	for (int r=0; r<dim; r++) {
		for (int c=0; c<dim; c++) {
			mult[(r * dim) + c] = matrix[(r * dim) + c] * n;
		}
	}
	return mult;
}

double * invert(int dim, double * matrix) {
	return multiply(dim, transposed(dim, comatrix(dim, matrix)), (1.0 / determinant(dim, matrix)));
}

void printMatrix(int dim, double * matrix) {
	for (int r=0; r<dim; r++) {
		fprintf(stdout, "| ");
		for (int c=0; c<dim; c++) {
			fprintf(stdout, "%lf ", matrix[(r * dim) + c]);
		}
		fprintf(stdout, " |\n");
	}
}

double * solveSystem(int dim, double * matrix, double * coeff) {
	double * result = (double *) malloc(dim * sizeof(double));
	double * inverted = invert(dim, matrix);

	if (DEBUG) {
		fprintf(stdout, "Inverted:\n");
		printMatrix(3, inverted);
	}

	for (int r=0; r<dim; r++) {
		result[r] = 0.0;
		for (int c=0; c<dim; c++) {
			result[r] += (inverted[(r * dim) + c] * coeff[c]);
		}
	}
	return result;
}

void printSystem(int dim, double * matrix, double * coeff) {
	char * unknowns = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
	for (int r=0; r<dim; r++) {
		for (int c=0; c<dim; c++) {
			fprintf(stdout, "%s(%lf x %c)", (c==0?"":" + "), matrix[(r * dim) +c], unknowns[c]);
		}
		fprintf(stdout, " = %lf\n", coeff[r]);
	}
}

int main(int argc, char **argv) {
	double matrix[9] = {
		 12,       13,      14,
		  1.345, -654,       0.001,
		 23.09,     5.3, -12.34
	};
	double constants[3] = { 234, 98.87, 9.876 };

	fprintf(stdout, "Solving system:\n");
	printSystem(3, &matrix[0], &constants[0]);
	double * result = solveSystem(3, &matrix[0], &constants[0]);
	fprintf(stdout, "A: %lf\n", result[0]);
	fprintf(stdout, "B: %lf\n", result[1]);
	fprintf(stdout, "C: %lf\n", result[2]);
}

File: docker/go/app/system.go
// Do see https://tour.golang.org/
package main

import (
	"fmt"
	"math"
)

func printMatrix(matrix [][]float64) {
	var size = len(matrix)
	var r, c int
	for r=0; r<size; r++ {
		fmt.Print("| ")
		for c=0; c<size; c++ {
			fmt.Printf(" %v", matrix[r][c])
		}
		fmt.Println(" |")
	}
}

func multiply(matrix [][]float64, n float64) [][]float64 {
	var res [][]float64
	// min is a slice
	var size = len(matrix)
	res = make([][]float64, size, size)
	var r, c int
	for r=0; r<size; r++ {
		res[r] = make([]float64, size, size)
	}

	for r=0; r<size; r++ {
		for c=0; c<size; c++ {
			res[r][c] = n * matrix[r][c]
		}
	}
	return res
}

func comatrix(matrix [][]float64) [][]float64 {
	var comat [][]float64
	// min is a slice
	var size = len(matrix)
	comat = make([][]float64, size, size)
	var r, c int
	for r=0; r<size; r++ {
		comat[r] = make([]float64, size, size)
	}

	for r=0; r<size; r++ {
		for c=0; c<size; c++ {
			comat[r][c] = determinant(minor(matrix, r, c)) * math.Pow(-1.0, float64(r + c + 2))
		}
	}
	return comat
}

func transposed(matrix [][]float64) [][]float64 {
	var transp [][]float64
	// min is a slice
	var size = len(matrix)
	transp = make([][]float64, size, size)
	var r, c int
	for r=0; r<size; r++ {
		transp[r] = make([]float64, size, size)
	}

	for r=0; r<size; r++ {
		for c=0; c<size; c++ {
			transp[r][c] = matrix[c][r]
		}
	}
	return transp
}

func invert(matrix [][]float64) [][]float64 {
	return multiply(transposed(comatrix(matrix)), (1.0 / determinant(matrix)))
}

func minor(matrix [][]float64, row int, col int) [][]float64 {

	var min [][]float64
	// min is a slice
	var size = len(matrix)
	min = make([][]float64, size - 1, size - 1)
	var r, c int
	for r=0; r<(size - 1); r++ {
		min[r] = make([]float64, size - 1, size - 1)
	}

	for r=0; r<size; r++ {
		if r != row {
			for c=0; c<size; c++ {
				if c != col {
					var mr, mc int
					if r < row {
						mr = r
					} else {
						mr = r - 1
					}
					if c < col {
						mc = c
					} else {
						mc = c - 1
					}
					min[mr][mc] = matrix[r][c]
				}
			}
		}
	}
	return min
}

func determinant(matrix [][]float64) float64 {
	var value float64 = 0

	if len(matrix) == 1 {
		value = matrix[0][0]
	} else {
		// C: column in Major
		var C int
		for C=0; C<len(matrix); C++ {
			var minDet = determinant(minor(matrix, 0, C))
			value += (matrix[0][C] * minDet * math.Pow(-1.0, float64(C + 1 + 1)))
		}
	}
	return value
}

func solveSystem(matrix [][]float64, coeffs []float64) []float64 {
	var size = len(coeffs)
	var result []float64
	result = make([]float64, size, size)
	var inverted = invert(matrix)
	var r, c int
	for r=0; r<size; r++ {
		result[r] = 0.0
		for c=0; c<size; c++ {
			result[r] += (inverted[r][c] * coeffs[c])
		}
	}
	return result
}

func printSystem(matrix [][]float64, coeffs []float64) {
	const UNKNOWNS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	var size = len(matrix)
	var r, c int
	for r=0; r<size; r++ {
		for c=0; c<size; c++ {
			if c > 0 {
				fmt.Print(" + ")
			}
			fmt.Printf("(%v x %c)", matrix[r][c], UNKNOWNS[c])
		}
		fmt.Printf(" = %v\n", coeffs[r])
	}
}

func main() {
	matrix := [][]float64 {
		{  12.0, 13.0, 14.0 },
		{ 1.345, -654, 0.001 },
		{ 23.09, 5.3, -12.34 }}

	coeffs := []float64 { 234, 98.87, 9.876 }

	fmt.Println("Resolving:")
	printSystem(matrix, coeffs)

	var result = solveSystem(matrix, coeffs)
	fmt.Printf("A = %v\n", result[0])
	fmt.Printf("B = %v\n", result[1])
	fmt.Printf("C = %v\n", result[2])
}
